### PR TITLE
Midnight delay option

### PR DESCRIPTION
**Pull request by hiqua**
#694

### CODE REVIEW

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/ButtonPanelView.kt [hunk 1]
```diff
@@ -54,6 +55,12 @@ abstract class ButtonPanelView<T : View>(
 
     @Synchronized
     protected fun inflateButtons() {
+        if (preferences.isMidnightDelayEnabled) {
+            setStartDayOffset(3, 0)
+        } else {
+            setStartDayOffset(0, 0)
+        }
+
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/ButtonPanelView.kt:63]: Is this required because the previous code only runs during startup? I think if we need this here, then we almost certainly need it in many other places (for example, in `HeaderView`). I guess this feature is more tricky than I expected. Perhaps we should force an app restart after the user changes this preference.

File: uhabits-android/src/main/java/org/isoron/uhabits/notifications/SnoozeDelayPickerActivity.kt [hunk 2]
```diff
@@ -49,7 +49,8 @@ class SnoozeDelayPickerActivity : FragmentActivity(), OnItemClickListener {
         val app = applicationContext as HabitsApplication
         val appComponent = app.component
         reminderController = appComponent.reminderController
-        habit = appComponent.habitList.getById(ContentUris.parseId(intent.data))
+        val data = intent.data
+        if (data != null) habit = appComponent.habitList.getById(ContentUris.parseId(data))
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/notifications/SnoozeDelayPickerActivity.kt:53]: Please revert if this is not directly related to the feature.

File: uhabits-android/src/main/java/org/isoron/uhabits/receivers/ReminderReceiver.kt [hunk 3]
```diff
@@ -46,7 +46,8 @@ class ReminderReceiver : BroadcastReceiver() {
         Log.i(TAG, String.format("Received intent: %s", intent.toString()))
         var habit: Habit? = null
         val today: Long = getStartOfTodayWithOffset()
-        if (intent.data != null) habit = habits.getById(ContentUris.parseId(intent.data))
+        val data = intent.data
+        if (data != null) habit = habits.getById(ContentUris.parseId(data))
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/receivers/ReminderReceiver.kt:50]: Please revert.

**Review by iSoron — CHANGES_REQUESTED** [review]
Thanks for the pull request, @hiqua. I just realized that this feature may be more tricky than expected. See comment below. Please let me know your thoughts.

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/ButtonPanelView.kt (hunk 1 above)
Comment (hiqua) [reply uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/ButtonPanelView.kt:63]: Ok then I think https://github.com/iSoron/uhabits/pull/739 might be more reasonable.

**Comment by iSoron** [comment]
Merged #739 instead of this.
